### PR TITLE
Create steam extract

### PR DESCRIPTION
**Pull request by Eibironke**
Be kind

### CODE REVIEW

File: pipeline/steam_extract.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+from os import environ as ENV
+from datetime import datetime, date
+from time import sleep
+from dotenv import load_dotenv
+import requests as req
+from bs4 import BeautifulSoup
+
+
+def get_rating(game_soup: BeautifulSoup) -> float:
```
Comment (hhm970) [inline pipeline/steam_extract.py:11]: Good use of type hints! Clearly documents your code

File: pipeline/steam_extract.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        if negative:
+            n_num = int(negative.find(
+                'span', class_='user_reviews_count').text.strip('()'))
+        else:
+            return 100
+
+        return round((p_num/(p_num+n_num)) * 100, 2)
+
+
+def get_platforms(game_soup: BeautifulSoup) -> list:
```
Comment (hhm970) [inline pipeline/steam_extract.py:35]: Your function appears to be able to return a `str` object, as well as a `list` object.

Make sure to reflect this in your type hints with

```
def get_platforms(game_soup: BeautifulSoup) -> list | str:
```

File: pipeline/steam_extract.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+    load_dotenv()
+
+    res = req.get(ENV["BASE_URL"], timeout=10)
+    soup = BeautifulSoup(res.text, features="html.parser")
+
+    all_containers = soup.find_all(
+        'a', class_="search_result_row")
+
+    print(get_everything(all_containers))
```
Comment (hhm970) [inline pipeline/steam_extract.py:171]: It's good practice to get rid of `print()` statements upon committing to GitHub

File: pipeline/steam_extract.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+
+def get_description(game_soup: BeautifulSoup) -> str:
+    """Function to get the description of a game from its URL"""
+    return game_soup.find('div', id='game_area_description').text.strip()
+
+
+def get_platform_ids(platform: list) -> list:
+    '''Returns a list of platform IDs given a
+    string of playable platforms.'''
```
Comment (hhm970) [inline pipeline/steam_extract.py:105]: Something I suggested to Shindy earlier today - for assigning platform names to their ids, we could use a global `dict` object assigning each platform name to its id?

So something like this:

```
{'Windows': 1, 'macOS': 2, 'Linux': 3}
```

File: pipeline/steam_extract.py (hunk 3 above)
Comment (Eibironke) [reply pipeline/steam_extract.py:171]: Will do!

File: pipeline/steam_extract.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
+"""Script to scrape relevant data from the epic games website"""
+
+from os import environ as ENV
+from datetime import datetime, date
+from time import sleep
+from dotenv import load_dotenv
+import requests as req
+from bs4 import BeautifulSoup
```
[the remark this answers is not in the document]
Comment (annalisev) [reply pipeline/steam_extract.py:8]: Separate out 3rd party imports.

File: pipeline/steam_extract.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+
+def get_rating(game_soup: BeautifulSoup) -> float:
+    """Function to get the rating percentage of a game from its URL"""
```
Comment (annalisev) [inline pipeline/steam_extract.py:12]:
```suggestion
    """Function to get the rating percentage of a game from its URL."""
```

File: pipeline/steam_extract.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
+"""Script to scrape relevant data from the epic games website"""
```
Comment (annalisev) [inline pipeline/steam_extract.py:1]:
```suggestion
"""Script to scrape relevant data from the epic games website."""
```

File: pipeline/steam_extract.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+            n_num = int(negative.find(
+                'span', class_='user_reviews_count').text.strip('()'))
+        else:
+            return 100
+
+        return round((p_num/(p_num+n_num)) * 100, 2)
+
+
+def get_platforms(game_soup: BeautifulSoup) -> list:
+    """Function to get the platforms a game is available on from its URL"""
```
Comment (annalisev) [inline pipeline/steam_extract.py:36]:
```suggestion
    """Function to get the platforms a game is available on from its URL."""
```

File: pipeline/steam_extract.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+            'div', class_="sysreq_contents")
+        platform_list = [platform.find(
+            'ul', class_='bb_ul').text for platform in platforms if platform.find(
+                'ul', class_='bb_ul')][0]
+        if "Windows" in platform_list:
+            return "Windows"
+        if "mac" in platform_list:
+            return "macOS"
+        if "Linux" in platform_list:
+            return "Linux"
```
Comment (annalisev) [inline pipeline/steam_extract.py]: Is it possible to have more that one platform returned?

File: pipeline/steam_extract.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+        return "Windows"
+
+
+def get_tags(game_soup: BeautifulSoup) -> list:
+    """Function to get the tags of a game from its URL"""
```
Comment (annalisev) [inline pipeline/steam_extract.py:63]:
```suggestion
    """Function to get the tags of a game from its URL."""
```

File: pipeline/steam_extract.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+            for i in game_soup.find_all('a', class_='app_tag')]
+
+
+def get_developer(game_soup: BeautifulSoup) -> str:
+    """Function to get the developing company of a game from its URL."""
+    return game_soup.find('div', class_='dev_row').text.split('\n')[-2]
+
+
+def get_publisher(game_soup: BeautifulSoup) -> str:
+    """Function to get the publisher of a game from its URL"""
```
Comment (annalisev) [inline pipeline/steam_extract.py:74]:
```suggestion
    """Function to get the publisher of a game from its URL."""
```

File: pipeline/steam_extract.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    pub_list = [x.text.split('\n') for x in pub]
+    for each in pub_list:
+        if 'Publisher:' in each:
+            return each[-2]
+    return None
+
+
+def get_name_price_date(container: BeautifulSoup) -> list:
+    """Function that returns a list containing the:
+      title, price and date of a game from its from the search page"""
```
Comment (annalisev) [inline pipeline/steam_extract.py:85]:
```suggestion
    """Function that returns a list containing the:
      title, price and date of a game from its from the search page."""
```

File: pipeline/steam_extract.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+    else:
+        float_price = float(price[1:])
+    release_date = container.find(
+        'div', {'class': "col search_released responsive_secondrow"}).text
+
+    return [title, float_price, release_date.strip()]
+
+
+def get_description(game_soup: BeautifulSoup) -> str:
+    """Function to get the description of a game from its URL"""
```
Comment (annalisev) [inline pipeline/steam_extract.py:99]:
```suggestion
    """Function to get the description of a game from its URL."""
```

File: pipeline/steam_extract.py [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+    if 'macOS' in platform:
+        id_list.append(2)
+    if 'Linux' in platform:
+        id_list.append(3)
+    return id_list
+
+
+def get_each_game_details(game_url: str) -> list:
+    """Function to get all the details of a game from its URL
+    returns a list"""
```
Comment (annalisev) [inline pipeline/steam_extract.py:118]:
```suggestion
    """Function to get all the details of a game from its URL
    returns a list."""
```

File: pipeline/steam_extract.py [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+    developer = get_developer(game_soup)
+    publisher = get_publisher(game_soup)
+    description = get_description(game_soup)
+
+    return [description, developer, publisher, date.today(), rating, 1, game_tags, platform_id_list]
+
+
+def get_everything(all_web_containers: BeautifulSoup) -> list[list]:
+    """Function to combine all the details of from the search results page
+      returns a list of lists"""
```
Comment (annalisev) [inline pipeline/steam_extract.py:135]:
```suggestion
    """Function to combine all the details of from the search results page
      returns a list of lists."""
```

File: pipeline/steam_extract.py [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+            for i in game_soup.find_all('a', class_='app_tag')]
+
+
+def get_developer(game_soup: BeautifulSoup) -> str:
+    """Function to get the developing company of a game from its URL"""
+    return game_soup.find('div', class_='dev_row').text.split('\n')[-2]
+
+
+def get_publisher(game_soup: BeautifulSoup) -> str:
+    """Function to get the publisher of a game from its URL"""
```
Comment (shindym) [inline pipeline/steam_extract.py]:
```suggestion
    """Function to get the publisher of a game from its URL."""
```

File: pipeline/steam_extract.py [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+    pub_list = [x.text.split('\n') for x in pub]
+    for each in pub_list:
+        if 'Publisher:' in each:
+            return each[-2]
+    return None
+
+
+def get_name_price_date(container: BeautifulSoup) -> list:
+    """Function that returns a list containing the:
+      title, price and date of a game from its from the search page"""
```
Comment (shindym) [inline pipeline/steam_extract.py]:
```suggestion
      title, price and date of a game from its from the search page."""
```

File: pipeline/steam_extract.py [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+    else:
+        float_price = float(price[1:])
+    release_date = container.find(
+        'div', {'class': "col search_released responsive_secondrow"}).text
+
+    return [title, float_price, release_date.strip()]
+
+
+def get_description(game_soup: BeautifulSoup) -> str:
+    """Function to get the description of a game from its URL"""
```
Comment (shindym) [inline pipeline/steam_extract.py]:
```suggestion
    """Function to get the description of a game from its URL."""
```

File: pipeline/steam_extract.py [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+        game_url = container['href']
+        try:
+            detail_list = get_each_game_details(game_url)
+            description = detail_list.pop(0)
+            name_price_date_list.append(name_price_date_list[1])
+            name_price_date_list[1] = description
+            final_list.append(name_price_date_list + detail_list)
+            sleep(1)
+        except IndexError:
+            continue
```
Comment (annalisev) [inline pipeline/steam_extract.py:156]: Is there a way to rework this section?

File: pipeline/steam_extract.py (hunk 9 above)
Comment (Eibironke) [reply pipeline/steam_extract.py]: The part above handles that so this just catches when theres only one platform

File: pipeline/steam_extract.py [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+def get_platform_ids(platform: list) -> list:
+    '''Returns a list of platform IDs given a
+    string of playable platforms.'''
+    id_list = []
+    if 'Windows' in platform:
+        id_list.append(1)
+    if 'macOS' in platform:
+        id_list.append(2)
+    if 'Linux' in platform:
+        id_list.append(3)
```
Comment (shindym) [inline pipeline/steam_extract.py:113]:
```suggestion
platform_id_translations = {'Windows': 1, 'macOS': 2, 'Linux':3}
for tag in platform:
        platform_id = platform_id_translations.get(tag, None)
        if platform_id is not None:
           id_list.append(platform_id)
```

Comment (shindym) [reply pipeline/steam_extract.py:113]: Following on from @hhm970 comment, this is how I did it.

File: pipeline/steam_extract.py [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+    platform_id_list = get_platform_ids(platform_list)
+    rating = get_rating(game_soup)
+    developer = get_developer(game_soup)
+    publisher = get_publisher(game_soup)
+    description = get_description(game_soup)
+
+    return [description, developer, publisher, date.today(), rating, 1, game_tags, platform_id_list]
+
+
+def get_everything(all_web_containers: BeautifulSoup) -> list[list]:
```
Comment (shindym) [inline pipeline/steam_extract.py:134]: Could rename to a better name - maybe `grab_all_games_details`?

File: pipeline/steam_extract.py [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+        name_price_date_list.pop(-1)
+        if game_date < date.today():
+            break
+        game_url = container['href']
+
+        detail_list = get_each_game_details(game_url)
+        description = detail_list.pop(0)
+        name_price_date_list.append(name_price_date_list[1])
+        name_price_date_list[1] = description
+        print(name_price_date_list + detail_list)
```
Comment (Eibironke) [inline pipeline/steam_extract.py:153]:
```suggestion
    
```

**Review by Eibironke — COMMENTED** [review]
👍

**Review by annalisev — DISMISSED** [review]
Changes look great!

**Pull request merged**